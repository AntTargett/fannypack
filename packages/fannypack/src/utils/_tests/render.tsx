import * as React from 'react';
import { render } from '@testing-library/react';
// @ts-ignore
import { Provider } from '../../Provider';

export default (Component: any, { theme = {} } = {}) => render(<Provider theme={theme}>{Component}</Provider>);
