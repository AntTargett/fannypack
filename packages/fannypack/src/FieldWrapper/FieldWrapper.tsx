import * as React from 'react';
import { Box as ReakitBox } from 'reakit';

import { useClassName, createComponent, createElement, createHook, useUniqueId } from '../utils';
import { Box, BoxProps } from '../Box';

import * as styles from './styles';

export type LocalFieldWrapperProps = {
  a11yId?: string;
  children:
    | (({ elementProps }: { elementProps: FieldElementProps }) => React.ReactNode)
    | React.ReactElement<any>
    | string;
  className?: string;
  description?: string | React.ReactElement<any>;
  hint?: string | React.ReactElement<any>;
  isOptional?: boolean;
  isRequired?: boolean;
  label?: string | React.ReactElement<any>;
  state?: string;
  tooltip?: string | React.ReactElement<any>;
  tooltipTrigger?: React.ReactElement<any>;
  validationText?: string;
};
export type FieldWrapperProps = BoxProps & LocalFieldWrapperProps;
export type FieldElementProps = {
  a11yId?: LocalFieldWrapperProps['a11yId'];
  isRequired?: LocalFieldWrapperProps['isRequired'];
  state?: LocalFieldWrapperProps['state'];
  marginTop?: string;
};

const useProps = createHook<FieldWrapperProps>(
  (props, themeKey) => {
    const { children, label } = props;

    const boxProps = Box.useProps(props);

    const className = useClassName({
      style: styles.FieldWrapper,
      styleProps: props,
      themeKey,
      prevClassName: boxProps.className
    });

    const uid = useUniqueId('FieldWrapper');

    return {
      ...boxProps,
      className,
      children: (
        <React.Fragment>
          {label && (
            <Box marginBottom="minor-2">
              <Box display="flex" alignItems="center" lineHeight="1">
                {typeof label === 'string' ? (
                  // @ts-ignore
                  <Box use="label" htmlFor={uid}>
                    {label}
                  </Box>
                ) : (
                  label
                )}
              </Box>
            </Box>
          )}
          {children}
        </React.Fragment>
      )
    };
  },
  { themeKey: 'FieldWrapper' }
);

export const FieldWrapper = createComponent<FieldWrapperProps>(
  props => {
    const textProps = useProps(props);
    return createElement({ children: props.children, component: ReakitBox, use: props.use, htmlProps: textProps });
  },
  {
    attach: {
      useProps
    },
    themeKey: 'FieldWrapper'
  }
);
