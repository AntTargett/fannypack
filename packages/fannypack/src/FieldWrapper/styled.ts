import { palette, theme } from 'styled-tools';
import Field from 'reakit/Field';

import styled, { space } from '../styled';
import { Omit } from '../types';
// @ts-ignore
import _Label from '../Label';
import Button from '../Button';
import Popover from '../Popover';
// @ts-ignore
import _Text from '../Text';
import { FieldWrapperProps } from './FieldWrapper';

export const Label = styled(_Label)`
  margin-bottom: 0 !important;

  & {
    ${theme('fannypack.FieldWrapper.label')};
  }
`;

export const DescriptionText = styled(_Text)`
  display: block;
  font-size: 0.8rem;

  & {
    ${theme('fannypack.FieldWrapper.description')};
  }
`;

export const HintText = styled(_Text)`
  display: block;
  font-size: 0.8rem;
  margin-top: ${space(1)}rem;

  & {
    ${theme('fannypack.FieldWrapper.hint')};
  }
`;

export const OptionalText = styled(_Text)`
  font-size: 0.8rem;
  color: ${palette('text100')};
  margin-left: ${space(2)}rem;
  line-height: 1;

  & {
    ${theme('fannypack.FieldWrapper.optional')};
  }
`;

export const RequiredText = styled(_Text)`
  color: ${palette('danger')};
  margin-left: ${space(1)}rem;
  font-weight: ${theme('fannypack.fontWeights.semibold')};
  font-size: ${theme('fannypack.fontSizes.150')}rem;

  & {
    ${theme('fannypack.FieldWrapper.required')};
  }
`;

export const TooltipButton = styled(Button)`
  margin-left: ${space(1)}rem;
  min-height: unset;
  height: 14px;
  padding: 0.1em 0.5em;

  & {
    ${theme('fannypack.FieldWrapper.TooltipButton.base')};
  }
`;

export const TooltipPopover = styled(Popover)`
  padding: ${space(1, 'major')}rem;
  width: max-content;
  max-width: 400px;

  & {
    ${theme('fannypack.FieldWrapper.TooltipPopover.base')};
  }
`;

export const ValidationText = styled(_Text)`
  display: block;
  font-size: 0.8rem;
  margin-top: ${space(1)}rem;

  & {
    ${theme('fannypack.FieldWrapper.validation')};
  }
`;

export const FieldWrapper = styled(Field)<Omit<FieldWrapperProps, 'children'>>`
  & {
    ${theme('fannypack.FieldWrapper.base')};
  }
`;

export default FieldWrapper;
