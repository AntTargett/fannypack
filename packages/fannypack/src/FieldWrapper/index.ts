import FieldWrapper from './FieldWrapper';
import * as styledFieldWrapper from './styled';

export default FieldWrapper;

export * from './FieldWrapper';
export { default as FieldWrapper } from './FieldWrapper';

export { styledFieldWrapper };
