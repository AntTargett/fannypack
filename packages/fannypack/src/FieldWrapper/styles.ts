import { cssClass } from '../styled';
import { theme } from '../utils';

export const FieldWrapper = styleProps => cssClass`
  & {
    ${theme(`${styleProps.themeKey}.base`)(styleProps)};
  }
`;
