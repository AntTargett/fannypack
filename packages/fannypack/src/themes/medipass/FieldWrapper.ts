import { css } from '../../styled';

export default {
  TooltipPopover: {
    base: css`
      background-color: rgb(255, 251, 217);
      line-height: 1.2;
    `
  }
};
