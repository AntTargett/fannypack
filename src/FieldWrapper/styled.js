// @flow
import styled, { css } from 'reakit/styled';
import { palette, theme } from 'styled-tools';
import Field from 'reakit/Field';

import _Label from '../Label';
import _Text from '../Text';

export const Label = styled(_Label)`
  margin-bottom: 0 !important;

  & {
    ${theme('fannypack.FieldWrapper.label')};
  }
`;

export const DescriptionText = styled(_Text)`
  display: block;
  font-size: 0.8rem;

  & {
    ${theme('fannypack.FieldWrapper.description')};
  }
`;

export const HintText = styled(_Text)`
  display: block;
  font-size: 0.8rem;
  margin-top: 0.25rem;

  & {
    ${theme('fannypack.FieldWrapper.hint')};
  }
`;

export const OptionalText = styled(_Text)`
  font-size: 0.8rem;
  color: ${palette('textLightest')};
  margin-left: 0.5rem;

  & {
    ${theme('fannypack.FieldWrapper.optional')};
  }
`;

export const ValidationText = styled(_Text)`
  display: block;
  font-size: 0.8rem;
  margin-top: 0.25rem;

  & {
    ${theme('fannypack.FieldWrapper.validation')};
  }
`;

export default styled(Field)`
  & {
    ${props =>
      props.isFullWidth &&
      css`
        width: 100%;
      `};
  }

  ${theme('fannypack.FieldWrapper.base')};
`;
