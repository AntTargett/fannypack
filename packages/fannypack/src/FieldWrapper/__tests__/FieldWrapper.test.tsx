import * as React from 'react';
import { renderHook } from '@testing-library/react-hooks';
import { FieldWrapper } from '../FieldWrapper';
import render from '../../utils/_tests/render';

describe('props', () => {
  it('should render correctly', () => {
    const { container } = render(<FieldWrapper>Hello world</FieldWrapper>);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render correctly with CSS props', () => {
    const { container } = render(<FieldWrapper color="primary">Hello world</FieldWrapper>);
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe('composition', () => {
  describe('as', () => {
    it('should render correctly', () => {
      const { container } = render(<FieldWrapper use="p">Hello world</FieldWrapper>);
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('hook', () => {
    it('should return with FieldWrapper props', () => {
      const { result } = renderHook(() => FieldWrapper.useProps());
      expect(result.current).toMatchSnapshot();
    });
  });

  describe('render props', () => {
    it('should render correctly', () => {
      const { container } = render(
        <FieldWrapper>{FieldWrapperProps => <div {...FieldWrapperProps}>Hello world</div>}</FieldWrapper>
      );
      expect(container.firstChild).toMatchSnapshot();
    });
  });
});

describe('theming', () => {
  it('FieldWrapper.base should render correctly', () => {
    const { container } = render(<FieldWrapper>hello world</FieldWrapper>, {
      // @ts-ignore
      theme: { FieldWrapper: { base: { backgroundColor: 'red' } } }
    });
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe('defaultProps', () => {
  it('should render correctly for className', () => {
    const { container } = render(<FieldWrapper>hello world</FieldWrapper>, {
      // @ts-ignore
      theme: { FieldWrapper: { defaultProps: { className: 'test', color: 'primary' } } }
    });
    expect(container.firstChild).toMatchSnapshot();
  });
});
