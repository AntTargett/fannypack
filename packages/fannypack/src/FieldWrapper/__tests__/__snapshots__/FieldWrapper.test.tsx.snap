// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`composition as should render correctly 1`] = `
.emotion-1 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-1:focus:not(:focus-visible) {
  outline: none;
}

<p
  class="emotion-0 emotion-1 emotion-2 fp-Box fp-FieldWrapper"
>
  Hello world
</p>
`;

exports[`composition hook should return with FieldWrapper props 1`] = `
Object {
  "children": <React.Fragment />,
  "className": "css-13h683e css-1pwrqjj css-3ekj6a fp-Box fp-FieldWrapper",
}
`;

exports[`composition render props should render correctly 1`] = `
.emotion-1 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-1:focus:not(:focus-visible) {
  outline: none;
}

<div
  class="emotion-0 emotion-1 emotion-2 fp-Box fp-FieldWrapper"
>
  Hello world
</div>
`;

exports[`defaultProps should render correctly for className 1`] = `
.emotion-1 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-1:focus:not(:focus-visible) {
  outline: none;
}

.emotion-2 {
  color: #3722ed !important;
}

<div
  class="emotion-0 emotion-1 emotion-2 test fp-Box fp-FieldWrapper"
>
  hello world
</div>
`;

exports[`props should render correctly 1`] = `
.emotion-1 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-1:focus:not(:focus-visible) {
  outline: none;
}

<div
  class="emotion-0 emotion-1 emotion-2 fp-Box fp-FieldWrapper"
>
  Hello world
</div>
`;

exports[`props should render correctly with CSS props 1`] = `
.emotion-1 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-1:focus:not(:focus-visible) {
  outline: none;
}

.emotion-2 {
  color: #3722ed !important;
}

<div
  class="emotion-0 emotion-1 emotion-2 fp-Box fp-FieldWrapper"
>
  Hello world
</div>
`;

exports[`theming FieldWrapper.base should render correctly 1`] = `
.emotion-1 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-1:focus:not(:focus-visible) {
  outline: none;
}

.emotion-0 {
  background-color: red;
}

<div
  class="emotion-0 emotion-1 emotion-2 fp-Box fp-FieldWrapper"
>
  hello world
</div>
`;
